pyshepseg/tiling.py

```diff
@@ -738,6 +744,332 @@ def crossesMidline(overlap, segLoc, orientation):
 
 
+def calcPerSegmentStatsTiled(imgfile, imgbandnum, segfile, maxSegId, 
+            statsSelection, chunkSize=DFLT_CHUNKSIZE):
+    """
+    Calculate selected per-segment statistics for the given band 
+    of the imgfile, against the given segment raster file. 
+    Calculated statistics are written to the segfile raster 
+    attribute table (RAT), so this file format must support RATs. 
+    
+    Calculations are carried out in a memory-efficient way, allowing 
+    very large rasters to be processed. Raster data is handled in 
+    small tiles, attribute table is handled in fixed-size chunks. 
+    
+    The statsSelection parameter is a list of tuples, one for each
+    statistics to be included. Each tuple is either 2 or 3 elements,
+        (columnName, statName) or (columnName, statName, parameter)
+    The 3-element form is used for any statistic which requires
+    a parameter, which currently is only the percentile. 
+    
+    The columnName is a string, used to name the column in the 
+    output RAT. 
+    The statName is a string used to identify which statistic 
+    is to be calculated. Available options are:
+        'min', 'max', 'mean', 'stddev', 'median', 'mode', 'percentile'.
+    The 'percentile' statistic requires the 3-element form, with 
+    the 3rd element being the percentile to be calculated. 
+    
+    For example
+        [('Band1_Mean', 'mean'),
+         ('Band1_stdDev', 'stddev'),
+         ('Band1_LQ', 'percentile', 25),
+         ('Band1_UQ', 'percentile', 75)
+        ]
+    would create 4 columns, for the per-segment mean and 
+    standard deviation of the given band, and the lower and upper 
+    quartiles, with corresponding column names. 
+
+    """
+    # Open the segment file
+    if isinstance(segfile, gdal.Dataset):
+        segds = segfile
+    else:
+        segds = gdal.Open(segfile, gdal.GA_Update)
+    segband = segds.GetRasterBand(1)
+    
+    # open the data file
+    if isinstance(imgfile, gdal.Dataset):
+        imgds = imgfile
+    else:
+        imgds = gdal.Open(imgfile, gdal.GA_Update)
+    imgband = imgds.GetRasterBand(imgbandnum)
+    
+    # Note that we skip the null segment ID, no stats are created for that. 
+    chunkMinVal = shepseg.MINSEGID
+
+    attrTbl = segband.GetDefaultRAT()
+    # Create columns, as required
+    for selection in statsSelection:
+        (colName, statName) = selection[:2]
+        colType = gdal.GFT_Integer
+        if statName in ('mean', 'stddev'):
+            colType = gdal.GFT_Real
+        attrTbl.CreateColumn(colName, colType, gdal.GFU_Generic)
+
+    while chunkMinVal <= maxSegId:
+        # This is one more than the largest seg id in the chunk
+        chunkMaxVal = chunkMinVal + chunkSize
+        if chunkMaxVal > maxSegId:
+            chunkMaxVal = maxSegId + 1
+
+        # Create per-segment histograms for current chunk
+        chunkCounts = calcCountsForChunk(segband, imgband, 
+                chunkMinVal, chunkMaxVal, attrTbl)
```

Comment:
Seems to me like we might as well write out the histo while we are here...

Reply:
I am not convinced. Remember, this is happening multiple times, for every band to be included in the stats, but we would only want to write the histogram once, so we would need some extra logic to control that. The counts which have been calculated here are counts of each different imagery value for the band in question, for each segment. I feel like it is simpler to keep that separate.